src/lobpcg/eig.rs

```diff
@@ -0,0 +1,268 @@
+//! Truncated eigenvalue decomposition
+//!
+use super::random;
+use crate::lobpcg::{lobpcg, LobpcgResult, TruncatedOrder as Order};
+
+use ndarray::prelude::*;
+use ndarray::{stack, NdFloat};
+use num_traits::{Float, NumCast};
+use std::iter::Sum;
+
+/// Truncated eigenproblem solver
+///
+/// This struct wraps the LOBPCG algorithm and provides convenient builder-pattern access to
+/// parameter like maximal iteration, precision and constraint matrix. Furthermore it allows
+/// conversion into a iterative solver where each iteration step yields a new eigenvalue/vector
+/// pair.
+///
+/// # Example
+///
+/// ```rust
+/// use ndarray::{arr1, Array2};
+/// use ndarray_linalg_rs::lobpcg::{TruncatedEig, TruncatedOrder};
+///
+/// let diag = arr1(&[1., 2., 3., 4., 5.]);
+/// let a = Array2::from_diag(&diag);
+///
+/// let eig = TruncatedEig::new(a, TruncatedOrder::Largest)
+///    .precision(1e-5)
+///    .maxiter(500);
+///
+/// let res = eig.decompose(3);
+/// ```
+pub struct TruncatedEig<A: NdFloat> {
+    order: Order,
+    problem: Array2<A>,
+    pub constraints: Option<Array2<A>>,
+    preconditioner: Option<Array2<A>>,
+    precision: f32,
+    maxiter: usize,
+}
+
+impl<A: Float + NdFloat + PartialOrd + Default + Sum> TruncatedEig<A> {
+    /// Create a new truncated eigenproblem solver
+    ///
+    /// # Properties
+    /// * `problem`: problem matrix
+    /// * `order`: ordering of the eigenvalues with [TruncatedOrder](crate::TruncatedOrder)
+    pub fn new(problem: Array2<A>, order: Order) -> TruncatedEig<A> {
+        TruncatedEig {
+            precision: 1e-5,
+            maxiter: problem.len_of(Axis(0)) * 2,
+            preconditioner: None,
+            constraints: None,
+            order,
+            problem,
+        }
+    }
+
+    /// Set desired precision
+    ///
+    /// This argument specifies the desired precision, which is passed to the LOBPCG solver. It
+    /// controls at which point the opimization of each eigenvalue is stopped. The precision is
+    /// global and applied to all eigenvalues with respect to their L2 norm.
+    ///
+    /// If the precision can't be reached and the maximum number of iteration is reached, then an
+    /// error is returned in [LobpcgResult](crate::lobpcg::LobpcgResult).
+    pub fn precision(mut self, precision: f32) -> Self {
+        self.precision = precision;
+
```

Comment:
These spaces in the setters are unnecessary IMO.

Reply:
but makes reading the code a bit easier

Reply:
I personally don't see the difference because I can tell the method is a "builder" from looking at the signature, but if you believe it's more readable then we can leave it as is